src/codegen/codegen-match.lisp

```diff
@@ -0,0 +1,320 @@
+  (declare (type list bindings)
+           (type t types)
+           (values t &optional))
+
+  `(declare
+    (ignorable ,@(mapcar #'car bindings))
+    ,@(if (not settings:*emit-type-annotations*)
+          nil
+          (loop :for binding :in bindings
+                :for var := (car binding)
+                :for type :in types
+                :collect `(type ,type ,var)))))
+
+(defun codegen-cond-match-branch (code pattern match-var match-expr-type env)
+  "Generate code for a `cl:cond' branch."
+  (declare (type t code)
+           (type pattern:pattern pattern)
+           (type symbol match-var)
+           (type tc:ty match-expr-type)
+           (type tc:environment env)
+           (values t &optional))
+
+  (multiple-value-bind (pred bindings types)
+      (codegen-pattern:codegen-pattern pattern match-var match-expr-type env)
+
+    `(,pred
+      ,(if (null bindings)
+           code
+           `(let ,bindings
+              ,(codegen-binding-types bindings types)
+              ,code)))))
+
+(defun codegen-case-match-branch (code pattern match-var match-expr-type env)
+  "Generate code for a `cl:case' branch."
+  (declare (type t code)
+           (type pattern:pattern pattern)
+           (type symbol match-var)
+           (type tc:ty match-expr-type)
+           (type tc:environment env)
+           (values t &optional))
+
+  (multiple-value-bind (pred bindings types)
+      (codegen-pattern:codegen-pattern pattern match-var match-expr-type env)
+    (declare (ignore pred))
+
+    (cond
+      ;; Case #1:
+      ;;
+      ;; Emit a list of a single literal pattern to match on.
+      ((pattern:pattern-literal-p pattern)
+       `((,(pattern:pattern-literal-value pattern))
+         ,code))
+
+      ;; Case #2:
+      ;;
+      ;; Emit a list of a single enum symbol to match on.
+      ((pattern:pattern-constructor-p pattern)
+       (let* ((name (pattern:pattern-constructor-name pattern))
+              (entry (tc:lookup-constructor env name)))
+         `((,(tc:constructor-entry-compressed-repr entry))
+           ,code)))
+
+      ;; Case #3:
+      ;;
+      ;; Emit a variable or wildcard binding in a catch-all branch.
+      (t
+       `(otherwise
+         (let ,bindings
+           ,(codegen-binding-types bindings types)
+           ,code))))))
+
+(defun codegen-match-branch (code pattern match-var match-expr-type env jumptablep)
+  "Generate code for a match branch."
+  (declare (type t code)
+           (type pattern:pattern pattern)
+           (type symbol match-var)
+           (type tc:ty match-expr-type)
+           (type tc:environment env)
+           (type boolean jumptablep)
+           (values t &optional))
+
+  (funcall (if jumptablep
+               #'codegen-case-match-branch
+               #'codegen-cond-match-branch)
+           code
+           pattern
+           match-var
+           match-expr-type
+           env))
+
+(defun codegen-cond-match (branches fallbackp)
```

Comment:
In these two functions, can you document `fallbackp`? It's not clear if `fallbackp` means "generate a fallback" or "assume branches includes a fallback".